dataset : https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud

In [1]:
pip install imbalanced-learn

In [5]:
import pandas as pd
df=pd.read_csv('/content/creditcard.csv')
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [7]:
df.shape

(284807, 31)

In [8]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

dependent variable - Class output value

In [9]:
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

there is huge difference between class 0 and 1


Independent and dependent Features

In [11]:
X=df.drop('Class',axis=1)
y=df['Class']

##Logistic Regression-Cross validation like kFold and Hyperparameter tuning

Types of Cross-Validation

Holdout Method:

The dataset is split into two parts: a training set and a testing set.
The model is trained on the training set and evaluated on the testing set.
Simple and quick, but the evaluation might depend heavily on how the split was made.

K-Fold Cross-Validation:
The dataset is divided into
𝑘
k equally sized folds.
The model is trained
𝑘
k times, each time using
𝑘
−
1
k−1 folds for training and the remaining fold for testing.
The final performance is the average performance across all
𝑘
k folds.
Helps in reducing the variance of the performance estimate.

Stratified K-Fold Cross-Validation:

Similar to K-Fold Cross-Validation, but ensures that each fold has a representative proportion of each class.
Useful for imbalanced datasets where some classes are underrepresented.

Leave-One-Out Cross-Validation (LOOCV):

Each instance in the dataset is used once as a test set while the remaining instances form the training set.
If there are
𝑛
n instances, the model is trained
𝑛
n times.
Provides a nearly unbiased estimate of model performance but is computationally expensive for large datasets.

Leave-P-Out Cross-Validation:

Similar to LOOCV, but instead of leaving one instance out,
𝑝
p instances are left out for the test set.
The model is trained on the remaining instances.
More flexible than LOOCV but computationally more expensive as
𝑝
p increases.

Time Series Cross-Validation:

Used for time series data where the order of data points is important.
The dataset is split in a way that respects the temporal order.
One common method is to use a rolling forecast origin.

Stratified Cross-Validation
Stratified K-Fold Cross-Validation ensures that each fold of the dataset has the same proportion of classes as the original dataset. This is particularly useful when dealing with imbalanced datasets. For example, if you have a binary classification problem with 90% of instances in class 0 and 10% in class 1, a simple K-Fold Cross-Validation might result in some folds having no instances of the minority class. Stratification helps to maintain the class distribution across folds.

Hyperparameter tuning is the process of optimizing the hyperparameters of a machine learning model to improve its performance. Hyperparameters are parameters that are not learned by the model from the data, but instead are set before the learning process begins. These parameters control the behavior of the training algorithm and the structure of the model. Tuning these parameters can significantly affect the performance of the model.

Key Concepts
Hyperparameters vs. Parameters:

Parameters: Values that are learned from the training data, such as weights in a neural network.
Hyperparameters: Values set before the learning process begins, such as learning rate, number of trees in a random forest, or the regularization parameter in logistic regression.
Common Hyperparameters:

Learning rate: Controls how much the model's weights are updated with respect to the loss gradient.
Batch size: Number of training samples used in one iteration.
Number of epochs: Number of times the entire training dataset is passed through the model.
Number of layers and nodes in neural networks.
Regularization parameters: Such as L1 and L2 regularization.
here we have used Grid Hypertuning
Exhaustively searches through a manually specified subset of the hyperparameter space.
Example: If tuning a learning rate (0.01, 0.1, 1) and batch size (32, 64, 128), grid search will evaluate the model with all combinations (3x3 = 9)

In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV
import numpy as np


In [16]:
10.0**np.arange(-2,3)

array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02])

In [25]:
log_class=LogisticRegression()
grid={'C':10.0**np.arange(-2,3),'penalty':['l1','l2']}
cv=KFold(n_splits=5,shuffle=False,random_state=None)

In [26]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.7,random_state=0)

In [27]:
clf=GridSearchCV(log_class,grid,cv=cv,n_jobs=-1,scoring='f1_macro')
clf.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py", line 1162, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py", line 54, in _check_solver
    raise ValueError(
ValueError: Solve

GridSearchCV(cv=KFold(n_splits=5, random_state=None, shuffle=False),
             estimator=LogisticRegression(), n_jobs=-1,
             param_grid={'C': array([1.e-02, 1.e-01, 1.e+00, 1.e+01, 1.e+02]),
                         'penalty': ['l1', 'l2']},
             scoring='f1_macro')

In [28]:
y=pred=clf.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[198935     93]
 [    94    243]]
0.9990620219195947
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199028
           1       0.72      0.72      0.72       337

    accuracy                           1.00    199365
   macro avg       0.86      0.86      0.86    199365
weighted avg       1.00      1.00      1.00    199365



Accuracy here is 99%
Confusion Matrix:
True Negatives (TN): 198935 instances were correctly predicted as negative.
False Positives (FP): 93 instances were incorrectly predicted as positive.
False Negatives (FN): 94 instances were incorrectly predicted as negative.
True Positives (TP): 243 instances were correctly predicted as positive.
precision:the ratio of true positives to the sum of true positives and false positives. It indicates the accuracy of the positive predictions.

Class 0: 1.0 (100% of the instances predicted as class 0 were actually class 0)
Class 1: 0.72 (72% of the instances predicted as class 1 were actually class 1)
F1-Score: The harmonic mean of precision and recall, providing a single measure of the model's performance for each class.

Class 0: 1.0
Class 1: 0.72

Always try to improve precision and recall and decrease FP and FN


#Random Forest

In [29]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train,y_train)

RandomForestClassifier()

In [30]:
y=pred=classifier.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[199009     19]
 [    87    250]]
0.9994683118902515
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199028
           1       0.93      0.74      0.83       337

    accuracy                           1.00    199365
   macro avg       0.96      0.87      0.91    199365
weighted avg       1.00      1.00      1.00    199365



fp decrease to 19 and fn to 87. precision 95%. recall is 79%. if we use hypertuning in random forest it will be better

#Random Forest with hypertuning

In [31]:
X_train.shape

(85442, 30)

In [32]:
y_train.value_counts()

Class
0    85287
1      155
Name: count, dtype: int64

weighing class 1 as 100 times here 155*100=15500 class 0 is 85287 itself for tuning. so that we are giving more importance to class 1 as it is less number

In [33]:
class_weight={0:1,1:100}

In [34]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier(class_weight=class_weight)
classifier.fit(X_train,y_train)

RandomForestClassifier(class_weight={0: 1, 1: 100})

In [35]:
y=pred=classifier.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[199011     17]
 [    91    246]]
0.9994582800391242
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    199028
           1       0.94      0.73      0.82       337

    accuracy                           1.00    199365
   macro avg       0.97      0.86      0.91    199365
weighted avg       1.00      1.00      1.00    199365



no much change from random forest without tuning


#Under Sampling

lot of disadvantage in under sampling because it will loss data. use only in small data.
here reduce points of maximum labels in class 0

In [36]:
y_train.value_counts()

Class
0    85287
1      155
Name: count, dtype: int64

In [38]:
from collections import Counter
Counter(y_train)

Counter({0: 85287, 1: 155})

In [51]:
from collections import Counter
Counter(y_train)
from imblearn.under_sampling import NearMiss
ns=NearMiss(sampling_strategy=0.8)
X_train_ns,y_train_ns=ns.fit_resample(X_train,y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))

The number of classes before fit Counter({0: 85287, 1: 155})
The number of classes after fit Counter({0: 193, 1: 155})


In [46]:
0.8*193


154.4

In [55]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)


RandomForestClassifier()

In [56]:
y=pred=classifier.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[152853  46175]
 [    13    324]]
0.7683244300654578
              precision    recall  f1-score   support

           0       1.00      0.77      0.87    199028
           1       0.01      0.96      0.01       337

    accuracy                           0.77    199365
   macro avg       0.50      0.86      0.44    199365
weighted avg       1.00      0.77      0.87    199365



so many errors and precision is very less

#Over Sampling

In [58]:
from imblearn.over_sampling import RandomOverSampler
os=NearMiss(sampling_strategy=0.5)
X_train_ns,y_train_ns=os.fit_resample(X_train,y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))

The number of classes before fit Counter({0: 85287, 1: 155})
The number of classes after fit Counter({0: 310, 1: 155})


In [59]:
from sklearn.ensemble import RandomForestClassifier
classifier=RandomForestClassifier()
classifier.fit(X_train_ns,y_train_ns)

RandomForestClassifier()

In [60]:
y=pred=classifier.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[190489   8539]
 [    32    305]]
0.9570085019938304
              precision    recall  f1-score   support

           0       1.00      0.96      0.98    199028
           1       0.03      0.91      0.07       337

    accuracy                           0.96    199365
   macro avg       0.52      0.93      0.52    199365
weighted avg       1.00      0.96      0.98    199365



#SMOTETomek

will be creating new points will be created. use both under and over sampling


In [61]:
from imblearn.combine import SMOTETomek

In [65]:
os=SMOTETomek(sampling_strategy=0.75)
X_train_ns,y_train_ns=os.fit_resample(X_train,y_train)
print('The number of classes before fit {}'.format(Counter(y_train)))
print('The number of classes after fit {}'.format(Counter(y_train_ns)))

The number of classes before fit Counter({0: 85287, 1: 155})
The number of classes after fit Counter({0: 84483, 1: 63161})


In [66]:
y=pred=classifier.predict(X_test)
print(confusion_matrix(y_test,pred))
print(accuracy_score(y_test,pred))
print(classification_report(y_test,pred))

[[190489   8539]
 [    32    305]]
0.9570085019938304
              precision    recall  f1-score   support

           0       1.00      0.96      0.98    199028
           1       0.03      0.91      0.07       337

    accuracy                           0.96    199365
   macro avg       0.52      0.93      0.52    199365
weighted avg       1.00      0.96      0.98    199365



#Ensemble Techniques

In [68]:
from imblearn.ensemble import EasyEnsembleClassifier

In [70]:
easy=EasyEnsembleClassifier()
easy.fit(X_train,y_train)

EasyEnsembleClassifier()

In [71]:
y_pred=easy.predict(X_test)
print(confusion_matrix(y_test,y_pred))
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))

[[192747   6281]
 [    36    301]]
0.9683143982143305
              precision    recall  f1-score   support

           0       1.00      0.97      0.98    199028
           1       0.05      0.89      0.09       337

    accuracy                           0.97    199365
   macro avg       0.52      0.93      0.54    199365
weighted avg       1.00      0.97      0.98    199365

